server/src/main/java/org/opensearch/cluster/routing/allocation/decider/ThrottlingAllocationDecider.java

```diff
@@ -150,36 +165,73 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
+            if(shardRouting.unassignedReasonIndexCreated()) {
+                return allocateInitialReplicas(shardRouting, node, allocation);
+            } else {
+                return allocateNonIntialReplicas(shardRouting, node, allocation);
+            }
+        }
+    }
+
+    private Decision allocateInitialReplicas(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        int currentInRecoveries = allocation.routingNodes().getInitialIncomingRecoveries(node.nodeId());
+        assert shardRouting.unassignedReasonIndexCreated() && !shardRouting.primary();
+
+        return allocateReplicas(shardRouting, allocation, currentInRecoveries, replicasInitialRecoveries,
+            (x,y) -> getInitialPrimaryNodeOutgoingRecoveries(x,y), replicasInitialRecoveries,
+            String.format(Locale.ROOT, "[%s=%d]", CLUSTER_ROUTING_ALLOCATION_NODE_INITIAL_REPLICAS_RECOVERIES_SETTING.getKey(),
+                replicasInitialRecoveries),
+            String.format(Locale.ROOT, "[%s=%d]", CLUSTER_ROUTING_ALLOCATION_NODE_INITIAL_REPLICAS_RECOVERIES_SETTING.getKey(),
+                replicasInitialRecoveries));
+    }
+
+    private Decision allocateNonIntialReplicas(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        int currentInRecoveries = allocation.routingNodes().getIncomingRecoveries(node.nodeId());
+
+        return allocateReplicas(shardRouting, allocation, currentInRecoveries, concurrentIncomingRecoveries,
+            (x,y) -> getPrimaryNodeOutgoingRecoveries(x,y), concurrentOutgoingRecoveries,
+            String.format(Locale.ROOT, "[%s=%d] (can also be set via [%s])",
+                CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_INCOMING_RECOVERIES_SETTING.getKey(),
+            concurrentIncomingRecoveries, CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey()),
+            String.format(Locale.ROOT, "[%s=%d] (can also be set via [%s])",
+                CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_OUTGOING_RECOVERIES_SETTING.getKey(),
+                concurrentOutgoingRecoveries, CLUSTER_ROUTING_ALLOCATION_NODE_CONCURRENT_RECOVERIES_SETTING.getKey()));
+    }
+
+    private Integer getPrimaryNodeOutgoingRecoveries(ShardRouting shardRouting, RoutingAllocation allocation) {
+        ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+        return allocation.routingNodes().getOutgoingRecoveries(primaryShard.currentNodeId());
+    }
+
+    private Integer getInitialPrimaryNodeOutgoingRecoveries(ShardRouting shardRouting, RoutingAllocation allocation) {
+        ShardRouting primaryShard = allocation.routingNodes().activePrimary(shardRouting.shardId());
+        return allocation.routingNodes().getInitialOutgoingRecoveries(primaryShard.currentNodeId());
+    }
+
+    private Decision allocateReplicas(ShardRouting shardRouting, RoutingAllocation allocation, int currentInRecoveries,
```

Comment:
Can this be peer primary recovery as well?

Reply:
yes.

Reply:
lets rename?

Reply:
okay.